pypesto/objective/petab.py

```diff
@@ -0,0 +1,108 @@
+"""Objective function for PEtab models using the PEtab simulator."""
+from __future__ import annotations
+
+try:
+    import petab.v1 as petab
+    from petab.v1.simulate import Simulator as PetabSimulator
+except ImportError:
+    petab = None
+from collections import OrderedDict
+from collections.abc import Sequence
+
+import numpy as np
+
+from ..C import FVAL, MODE_FUN, MODE_RES, RES, ModeType
+from .base import ObjectiveBase, ResultDict
+
+
+class PetabSimulatorObjective(ObjectiveBase):
+    """Objective function for PEtab models using the PEtab simulator."""
+
+    def __init__(
+        self,
+        simulator: PetabSimulator,
+        x_names: Sequence[str] | None = None,
+    ):
+        """Initialize the PEtab simulator objective function.
+
+        Parameters
+        ----------
+        petab_problem:
+            The PEtab problem.
+        simulator:
+            The PEtab simulator.
+        x_names:
+            Names of optimization parameters.
+        """
+        if petab is None:
+            raise ImportError(
+                "The `petab` package is required for this objective function."
+            )
+        self.simulator = simulator
+        self.petab_problem = self.simulator.petab_problem
+        if x_names is None:
+            x_names = list(self.petab_problem.get_x_ids())
+        super().__init__(x_names=x_names)
+
+    def replace_parameters(self, x: np.ndarray):
+        """Replace the parameters in the PEtab problem with the given values.
+
+        Parameters
+        ----------
+        x:
+            Parameter vector for optimization.
+        """
+        x_dict = OrderedDict(zip(self._x_names, x))
+        x_unscaled = self.petab_problem.unscale_parameters(x_dict)
+        par_df = self.petab_problem.parameter_df
+        par_df["nominalValue"] = par_df.index.map(x_unscaled)
+        # add here try except whether simulator has its own method for this
+        try:
+            self.simulator.set_parameters(x_unscaled)
+        except AttributeError:
+            self.simulator = self.simulator.__class__(
+                petab_problem=self.petab_problem
```

Comment:
@PaulJonasJost  The issue with the reinitialization here, is that any parameters set before using custom constructors will be lost. I'm not sure how bad that is, but at the very lest, the working directory should be passed along as well. 

just something to think about

Reply:
I think it is not feasible to accurately reinitialise, thus i removed the tryExcept and we expect the Simulator to have such a functionality.